### PR TITLE
Add support for options (ro/rw) in volumes_from (#1188)

### PR DESCRIPTION
**Pull request by vdemeester**
Add support for ``volumes_from`` with read-only, with integration and unit tests, related to #1188.

```yaml
container:
  build: .
  ports:
    - '19198:22'
  volumes_from:
    - 'nginx_logs_1:ro'
```

This changes a bit how ``Service`` treats its ``volumes_from`` attribute, as it now asks for a ``tuple`` of ``(object, string)`` where the object can be a ``Service`` or ``Container`` and the string could be ``None`` or a option like ``'ro'``, ``'rw'``.

I think there is rooms of improvement in the implementation, so reviews are more than welcome :wink:.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: compose/service.py [hunk 1]
```diff
@@ -334,16 +334,21 @@ def _get_links(self, link_to_self):
 
     def _get_volumes_from(self):
         volumes_from = []
-        for volume_source in self.volumes_from:
+        for volume_from in self.volumes_from:
+            volume_source, mount_option = volume_from
```
Comment (dnephin) [inline compose/service.py]: A namedtuple would be good, that way you could do `volume_from.source` and `volume_from.mount_option`.

**Comment by vdemeester** [comment]
@dnephin updated with a ``namedtuple`` (defined in ``config.py``, wasn't really sure about that..)

**Comment by funkyfuture** [comment]
i suggest to put the declaration of `VolumesFrom` into `__init.py__`. it's now placed in a module that itself doesn't make use of it, but various others. this would also prevent later possible circular references through imports.

**Comment by vdemeester** [comment]
@funkyfuture ah yes, good idea, updated my PR with that :wink:

**Comment by funkyfuture** [comment]
you may change the issue-title as not only `ro` is supported as suffix.

File: compose/__init__.py [hunk 2]
```diff
@@ -1,4 +1,8 @@
 from __future__ import unicode_literals
 from .service import Service  # noqa:flake8
 
+from collections import namedtuple
+
 __version__ = '1.3.0dev'
+
+VolumeFrom = namedtuple("VolumeFrom", 'source mount_option')
```
Comment (dnephin) [inline compose/__init__.py]: minor point, but this is only used in one module right now (outside of tests), so it would be good to keep it local to that module.  

It's good to keep `__init__.py` as small as possible

**Comment by dnephin** [comment]
I think this is looking good.

At some point we're going to need to refactor part of this. It seems like the logic overlaps a lot with `net` handling and is pretty similar between `config` and `service` modules, but that shouldn't happen as part of this change.

File: compose/__init__.py (hunk 2 above)
Comment (vdemeester) [inline compose/__init__.py]: Yeah that's right.. I moved it back to project.py — i'll see if janky is happy.. I'm on a EDGE network right now, it's taking for ever to build/test it on my laptop :cry: .

**Comment by vdemeester** [comment]
I've update and rebase it against master (as there was a small conflict with a test rename).

**Comment by funkyfuture** [comment]
well, it's not in `config.py` anymore. `project.py` seems more proper.

however, i learned yesterday that CPython 2.6 doesn't have `NamedTuple` in its library. but for some reason that interpreter is supposed to be supported.

**Comment by vdemeester** [comment]
@funkyfuture ah :disappointed:, not cool.. 

> With the 2.6.9 release, and five years after its first release, the Python 2.6 series is now officially retired. All official maintenance for Python 2.6, including security patches, has ended.

Are we sure we want to continue supporting a version that is not supported by the python foundation since 2 years ? :P

**Comment by aanand** [comment]
Looks like OS X has shipped with 2.7 since Mavericks. I think we can probably drop 2.6 support now.

**Comment by aanand** [comment]
I like the look of this. Just needs docs.

**Comment by vdemeester** [comment]
Oh yeah, I didn't think of updating documentation. I'll update this in a few hours ;-)

**Comment by dnephin** [comment]
>  CPython 2.6 doesn't have NamedTuple

Sure it does! https://docs.python.org/2.6/library/collections.html#collections.namedtuple

It was added in 2.6, we use it in a few places already in compose.

**Comment by vdemeester** [comment]
Documentation updated.

**Comment by funkyfuture** [comment]
> Sure it does!

sorry for the confusion and driving this off-topic. i wrongly assumed `sys.version_info` was a named tuple. so, thank you for the correction. however, dropping 2.6 seems cool.

File: tests/integration/project_test.py [hunk 3]
```diff
@@ -67,7 +87,30 @@ def test_volumes_from_container(self):
             client=self.client,
         )
         db = project.get_service('db')
-        self.assertEqual(db.volumes_from, [data_container])
+        self.assertEqual(db._get_volumes_from(), [data_container.id])
+
+        project.kill()
+        project.remove_stopped()
```
Comment (dnephin) [inline tests/integration/project_test.py]: These two lines wont run if the test fails. The test case should take care of this for you, so you can remove it.

File: tests/integration/project_test.py [hunk 4]
```diff
@@ -67,7 +87,30 @@ def test_volumes_from_container(self):
...
+                    'volumes_from': ['composetest_data_container:ro'],
+                },
+            }),
+            client=self.client,
+        )
+        db = project.get_service('db')
+        self.assertEqual(db._get_volumes_from(), [data_container.id + ':ro'])
 
         project.kill()
         project.remove_stopped()
```
Comment (dnephin) [inline tests/integration/project_test.py]: Same here, if you don't mind removing it, that would be great.

File: tests/integration/service_test.py [hunk 5]
```diff
@@ -236,8 +237,8 @@ def test_create_container_with_home_and_env_var_in_volume_path(self):
     def test_create_container_with_volumes_from(self):
         volume_service = self.create_service('data')
         volume_container_1 = volume_service.create_container()
-        volume_container_2 = Container.create(self.client, image='busybox:latest', command=["top"])
-        host_service = self.create_service('host', volumes_from=[volume_service, volume_container_2])
+        volume_container_2 = Container.create(self.client, image='busybox:latest', command=["/bin/sleep", "300"])
```
Comment (dnephin) [inline tests/integration/service_test.py]: I think this is a bad merge, it should be `top`

File: tests/integration/service_test.py [hunk 6]
```diff
@@ -245,6 +246,18 @@ def test_create_container_with_volumes_from(self):
         self.assertIn(volume_container_2.id,
                       host_container.get('HostConfig.VolumesFrom'))
 
+    def test_create_container_with_volumes_from_read_only(self):
+        volume_service = self.create_service('data')
+        volume_container_1 = volume_service.create_container()
+        volume_container_2 = Container.create(self.client, image='busybox:latest', command=["/bin/sleep", "300"])
```
Comment (dnephin) [inline tests/integration/service_test.py]: Please use `top`, we changed everything over to use it so the tests would run faster

**Comment by vdemeester** [comment]
@dnephin Updated :wink:

**Comment by aanand** [comment]
As @samalba reports in https://github.com/docker/compose/issues/1493, it'd be good to support all the modes the Docker daemon supports. This PR might already do that, but it'd be good to test and document it.

**Comment by vdemeester** [comment]
@aanand ah yes, I should update the documentation, and probably tests as well then :wink:

**Comment by vdemeester** [comment]
Well, it's just gonna work with Docker 1.7.0 as it's a pretty recent change (https://github.com/docker/docker/commit/af7d17a6c906d348eb0929766a81fd04fc173c44). 

@aanand How should we handle that ?

**Comment by vdemeester** [comment]
Few bugs on the build with 1.7.0-rc1, I'll wait for https://github.com/docker/compose/pull/1497 to be merged then :). I've update the Dockerfile to only use the 1.7.0-rc1, but I'll update this when needed :wink:.

**Comment by aanand** [comment]
I've implemented support for any mode with the `volumes` key in https://github.com/docker/compose/pull/1515, so here it just needs to be implemented for `volumes_from`.

**Comment by vdemeester** [comment]
@aanand ok :wink:. I think that's already what's implemented (I look at #1515 and this was nothing I modified :stuck_out_tongue:)

**Comment by aanand** [comment]
Great.

I think the dependency on Docker 1.7 can actually be relaxed, since my understanding is that the volume mode support might be backported to the 1.6 series.

This means that your integration tests won't pass, but as long as there's a test in `test/unit/service.py` that uses a mock client (like [test_create_with_special_volume_mode](https://github.com/docker/compose/pull/1515/files#diff-bca0836805752707d5852da846492291R395) in my PR) it should be fine to remove the integration tests.

**Comment by vdemeester** [comment]
I've rebased it since #1497 has been merged :wink:.

**Comment by vdemeester** [comment]
ohhh @aanand, just got what you meant… :sweat_smile: I'll update it this week end then :smile:

**Comment by vdemeester** [comment]
@aanand just rebased it (there was a conflict on the yaml file) and removed the integration tests as said.

**Comment by vdemeester** [comment]
@aanand @dnephin @bfirsh Anything more to do on this one ? :blush:

File: compose/project.py [hunk 7]
```diff
@@ -13,6 +14,8 @@
 
 log = logging.getLogger(__name__)
 
+VolumeFrom = namedtuple("VolumeFrom", 'source mount_option')
```
Comment (aanand) [inline compose/project.py]: For the sake of consistency, could you:

1. rename `mount_option` to `mode`
2. rename `VolumeFrom` to `VolumesFromSpec`
3. move it into `compose.service` and import it here

Comment (aanand) [inline compose/project.py]: (We should really move all those named tuples into their own file, along with any related functions for parsing them)

File: compose/project.py [hunk 8]
```diff
@@ -158,15 +164,19 @@ def get_volumes_from(self, service_dict):
         volumes_from = []
         if 'volumes_from' in service_dict:
             for volume_name in service_dict.get('volumes_from', []):
+                if ':' in volume_name:
+                    service_name, mount_option = volume_name.split(':', 1)
+                else:
+                    service_name, mount_option = volume_name, None
```
Comment (aanand) [inline compose/project.py]: Could you extract this logic into a function, and do the same with the string-building logic in `Service._get_volumes_from()`? Say, `parse_volumes_from()` and `build_volumes_from()`.

File: compose/project.py (hunk 7 above)
Comment (vdemeester) [inline compose/project.py]: Ok, will do :wink:.

Moving them in their own file along with related functions should be in another PR I think, right ?

File: compose/service.py [hunk 9]
```diff
@@ -442,16 +442,20 @@ def _get_links(self, link_to_self):
...
+            else:
+                mount_option = ''
+            if isinstance(volume_from.source, Service):
+                containers = volume_from.source.containers(stopped=True)
                 if not containers:
-                    volumes_from.append(volume_source.create_container().id)
+                    volumes_from.append(volume_from.source.create_container().id + mount_option)
                 else:
-                    volumes_from.extend(map(attrgetter('id'), containers))
+                    volumes_from.extend([container_id + mount_option for container_id in map(attrgetter('id'), containers)])
```
Comment (aanand) [inline compose/service.py]: Now that this is a comprehension, there's not much point in keeping the `map`+`attrgetter` stuff - it can just iterate over `containers`.

File: compose/project.py (hunk 7 above)
Comment (aanand) [inline compose/project.py]: Agreed.

File: compose/project.py (hunk 8 above)
Comment (vdemeester) [inline compose/project.py]: Ok.

**Comment by thaJeztah** [comment]
ping @aanand PTAL; looks like the PR was updated :)

**Comment by vdemeester** [comment]
@thaJeztah he did, he did ;-). I still have to update some code though :-).

**Comment by thaJeztah** [comment]
@vdemeester :+1: wasn't sure if you forgot to ping him :smile:

File: compose/service.py [hunk 10]
```diff
@@ -770,6 +765,47 @@ def parse_volume_spec(volume_config):
...
+        if not containers:
+            volumes_from = ["{}:{}".format(volume_from_spec.source.create_container().id, volume_from_spec.mode)]
+        else:
+            volumes_from = ["{}:{}".format(container.id, volume_from_spec.mode) for container in containers]
+    elif isinstance(volume_from_spec.source, Container):
+        volumes_from = ["{}:{}".format(volume_from_spec.source.id, volume_from_spec.mode)]
+    return volumes_from
+
+
+def parse_volume_from_spec(volume_from_config, get_service, client):
```
Comment (vdemeester) [inline compose/service.py]: I did something a little strange I feel, because I need to call ``get_service`` or use ``client`` to get the source (service or container). I ended up passing the method and the client reference as parameters.

File: compose/service.py [hunk 11]
```diff
@@ -857,3 +893,12 @@ def build_extra_hosts(extra_hosts_config):
         "extra_hosts_config \"%s\" must be either a list of strings or a string->string mapping," %
         extra_hosts_config
     )
+
+
+class NoSuchService(Exception):
```
Comment (vdemeester) [inline compose/service.py]: This one was defined in ``project.py`` so moved here to avoid circular dependency.

File: compose/service.py (hunk 10 above)
Comment (aanand) [inline compose/service.py]: Sorry, I was unclear. I was thinking that just the string splitting and concatenation logic should be moved into utility functions - the part where we determine whether the source is a service or a container (as well as the corresponding logic in `Service._get_volumes_from` where we check whether we're looking at an instance of Service or Container) can stay where it is.

Comment (vdemeester) [inline compose/service.py]: Don't be sorry :wink: ; that's what I understood, but… I though I could do both in there, to avoid updating VolumeFromSpec afterwards. But now that I'm looking at it, it feels really bad :sweat_smile:, I'll fix that :wink:.

File: compose/project.py [hunk 12]
```diff
@@ -157,16 +160,19 @@ def get_links(self, service_dict):
...
         if 'volumes_from' in service_dict:
-            for volume_name in service_dict.get('volumes_from', []):
+            for volume_from_config in service_dict.get('volumes_from', []):
+                volume_from_spec = parse_volume_from_spec(volume_from_config)
+                # Get service
                 try:
-                    service = self.get_service(volume_name)
-                    volumes_from.append(service)
+                    service_name = self.get_service(volume_from_spec.source)
+                    volume_from_spec = VolumeFromSpec(service_name, volume_from_spec.mode)
```
Comment (vdemeester) [inline compose/project.py]: @aanand I wasn't sure if ``parse_volume_from_spec`` should return a ``VolumeFromSpec`` as I need to create a new one few lines later..

**Comment by vdemeester** [comment]
@aanand I think i've update the code according to all your comments :)

**Comment by vdemeester** [comment]
@aanand rebased it (but seems like janky is not happy :sob:)

**Comment by thaJeztah** [comment]
kicked Janky (in case it helps :-))

**Comment by vdemeester** [comment]
@thaJeztah it did :wink:

**Comment by vdemeester** [comment]
ping @aanand @mnowster

File: compose/project.py [hunk 13]
```diff
@@ -24,12 +24,15 @@ def sort_service_dicts(services):
     def get_service_names(links):
         return [link.split(':')[0] for link in links]
 
+    def get_service_names_from_volumes_from(volumes_from):
+        return [volume_from.split(':')[0] for volume_from in volumes_from]
```
Comment (aanand) [inline compose/project.py]: Could we also use `parse_volume_from_spec` here? Like:

```python
return [
    parse_volume_from_spec(vf).source
    for vf in volumes_from
]
```

File: compose/service.py [hunk 14]
```diff
@@ -812,6 +807,38 @@ def parse_volume_spec(volume_config):
...
+
+    if len(parts) == 1:
+        source = parts[0]
+        mode = 'rw'
+    else:
+        source, mode = parts
+
+    if mode not in ('rw', 'ro'):
+        raise ConfigError("VolumeFrom %s has invalid mode (%s), should be "
+                          "one of: rw, ro." % (volume_from_config, mode))
```
Comment (aanand) [inline compose/service.py]: We shouldn't validate the mode client-side - we can leave that to the daemon.

**Comment by aanand** [comment]
After stepping back, I'm a little concerned that in various parts of the code, a `VolumeFromSpec`'s `source` property can be either a string or a service/container. I don't think we can escape the service/container conflation, but adding strings into the mix feels bad to me.

(I'm aware this was informed by my suggestions, which might not have been very good ones. Sorry.)

ping @dnephin @mnowster - what do you think of the code?

File: compose/project.py (hunk 13 above)
Comment (dnephin) [inline compose/project.py]: +1

**Comment by dnephin** [comment]
Sorry it's taken me a while to look at this.

I think overall it looks great, but I share the concern about using `VolumeFromSpec` with different types for source.  I'd even say let's create another namedtuple `Volume`, that will probably have the same fields, but `source` is expected to be a container or service, instead of a string.

That should make it a bit more clear, and maybe in the future those types will diverage.

**Comment by mnowster** [comment]
Hi there @vdemeester , would you have the time to address the feedback from @dnephin ?  We'd like to get this PR in for 1.5. 

If you don't have the time, no worries, I can pick this up(your contributions will remain in tact). Let me know, thanks :)

**Comment by vdemeester** [comment]
@mnowster waw it's been a while :sweat_smile:. There is kind of a lot of conflicts and my head is a little lost on `compose` on that one :sweat:, so feel free to *pick it up* :wink:.

**Comment by mnowster** [comment]
@vdemeester I know :disappointed:  I'm sorry it's taken so long, been so much going on but we're trying to get better at responding. No problem, I'll pick it up, thank you for all your work on this so far, your commits will remain so you'll be credited :+1: :smiley:
